userspace/libscap/engine/savefile/converter/converter.cpp

```diff
@@ -0,0 +1,407 @@
+	// previous syscall is already completed.
+
+	clear_evt(tid);
+
+	scap_evt *tmp_evt = (scap_evt *)malloc(evt->len);
+	if(!tmp_evt) {
+		throw std::runtime_error("Cannot allocate memory for the enter event.");
+	}
+	memcpy(tmp_evt, evt, evt->len);
+	evt_storage[tid] = safe_scap_evt(tmp_evt);
+}
+
+static scap_evt *retrieve_evt(uint64_t tid) {
+	if(evt_storage.find(tid) != evt_storage.end()) {
+		return evt_storage[tid].get();
+	}
+	return nullptr;
+}
+
+static uint16_t get_param_len(scap_evt *evt, uint8_t num_param) {
+	if(evt->nparams <= num_param) {
+		std::string error = "Try to access len of param num '" + std::to_string(num_param) +
+		                    "' for event " + get_event_name((ppm_event_code)evt->type) + "_" +
+		                    get_direction_char((ppm_event_code)evt->type) +
+		                    " (num parameters: " + std::to_string(evt->type) + ").";
+		throw std::runtime_error(error);
+	}
+
+	// todo!: we need to manage LARGE_PAYLOAD events
+	uint16_t off_len = sizeof(scap_evt) + sizeof(uint16_t) * num_param;
+	uint16_t len = 0;
+	memcpy(&len, (char *)evt + off_len, sizeof(uint16_t));
+	return (uint32_t)len;
+}
+
+static char *get_param_ptr(scap_evt *evt, uint8_t num_param) {
+	if(evt->nparams <= num_param) {
+		std::string error = "Try to access param num '" + std::to_string(num_param) +
+		                    "' for event " + get_event_name((ppm_event_code)evt->type) + "_" +
+		                    get_direction_char((ppm_event_code)evt->type) +
+		                    " (num parameters: " + std::to_string(evt->type) + ").";
+		throw std::runtime_error(error);
+	}
+
+	char *ptr = (char *)evt + sizeof(scap_evt) + sizeof(uint16_t) * evt->nparams;
+	uint16_t ptr_off = 0;
+	for(auto i = 0; i < num_param; i++) {
+		uint16_t len = 0;
+		memcpy(&len, (char *)evt + sizeof(scap_evt) + sizeof(uint16_t) * i, sizeof(uint16_t));
+		ptr_off += len;
+	}
+
+	return ptr + ptr_off;
+}
+
+// This writes len + the param
+static void push_default_parameter(scap_evt *evt, uint16_t *params_offset, uint8_t param_num) {
+	// Please ensure that `new_evt->type` is already the final type you want to obtain.
+	// Otherwise we will access the wrong entry in the event table.
+	const struct ppm_event_info *event_info = &(g_event_info[evt->type]);
+	uint16_t len = scap_get_size_bytes_from_type(event_info->params[param_num].type);
+	uint16_t lens_offset = sizeof(scap_evt) + param_num * sizeof(uint16_t);
+
+	PRINT_MESSAGE(
+	        "push default param (%d, type: %d) with len (%d) at {params_offest (%d), "
+	        "lens_offset (%d)}\n",
+	        param_num,
+	        event_info->params[param_num].type,
+	        len,
+	        *params_offset,
+	        lens_offset);
+
+	// The default param will be always 0 so we just need to copy the right number of 0 bytes.
+	// `uint64_t` should be enough for all the types considering that types like CHARBUF, BYTEBUF
+	// have `len==0`
+	static uint64_t val = 0;
```

Comment:
```suggestion
	uint64_t val = 0;
```

No need for static here, right?

Reply:
we could keep it if we don't want to allocate it at each conversion but probably it changes almost nothing